feat: support multi-document values files

LuBingtan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
close #13097 
**Special notes for your reviewer**:

LuBingtan committed:
feat: support multi-document values files

Signed-off-by: lubingtan <bingtanlu@gmail.com>

robertsirc commented:
Hello, thank you for your pull request. This function doesn't have any associated unit test for this. Is it possible to have that added to ensure we don't break functionality.

gjenkins8 reviewed (changes requested):
Along with @robertsirc comment about tests, this only addresses values files specified on the command line. I think for consistency, we should support multi-document values files for both cli overrides, and default chart values.

This also somewhat implies we should unify the value file load process between chart's and the cli.

(it could be argued we do one first, but at least my opinion is we should do both at thg same time for consistency)

LuBingtan commented:
Thanks for your comments! Let me update

LuBingtan committed:
feat: support multi-document values files for default chart values

Signed-off-by: lubingtan <bingtanlu@gmail.com>

LuBingtan commented:
Hi @robertsirc @gjenkins8 ,
I've updated accordingly, and let me exaplain briefly what I did here:

1. Added a public function `LoadValues` along with its UT under `pkg/chart/loader/load.go`. This function can be used by both default chart values and the CLI.
2. To facilitate the use of the function `mergeMaps`, I also relocated it and its UT to `pkg/chart/loader/load.go`.

Not sure if this change is optimal. Please take a look, would appreciate your feedback.

gjenkins8 commented on pkg/chart/loader/load.go:
```diff
@@ -205,3 +206,45 @@ func LoadFiles(files []*BufferedFile) (*chart.Chart, error) {
 
 	return c, nil
 }
+
```
Can we add a doc string please

gjenkins8 commented on pkg/chart/loader/load.go:
```diff
@@ -205,3 +206,45 @@ func LoadFiles(files []*BufferedFile) (*chart.Chart, error) {
 	return c, nil
 }
+
+func LoadValues(data []byte) (map[string]interface{}, error) {
```
```suggestion
func LoadValues(data io.Reader) (map[string]interface{}, error) {
```

It is preferred to take an interface that can support the necessary functionality, rather than a concrete type

gjenkins8 commented on pkg/chart/loader/load_test.go:
```diff
@@ -488,6 +489,118 @@ func TestLoadInvalidArchive(t *testing.T) {
 }
 
+func TestLoadValues(t *testing.T) {
+	testDatas := []struct {
```
nit: `testCases` might be better (and more consistent) naming/terminology here

```suggestion
	testCases := []struct {
```

gjenkins8 replied:
Personally, I prefer:
```
testCases := map[string]struct {
```

Where the map key is the name of the test. It does make the test ordering non-deterministic (which IMHO is a good things). Up to you if you would also prefer a map here.

gjenkins8 reviewed (commented):
Nice, thanks for the update. Close to LGTM, few minor comments

LuBingtan committed:
fix: use Reader interface as the input of LoadValues and enhance UT of LoadValues

Signed-off-by: lubingtan <bingtanlu@gmail.com>

LuBingtan commented:
Hi @gjenkins8  thanks for your comments, I've updated accordingly, please take a look again

TerryHowe commented on pkg/chart/loader/load.go:
```diff
@@ -205,3 +206,46 @@ func LoadFiles(files []*BufferedFile) (*chart.Chart, error) {
+	}
+	return values, nil
+}
+
```
This should technically have a doc string too since it is exported.

LuBingtan replied:
thanks, also updated here

LuBingtan committed:
fix: add doc for func MergeMaps

Signed-off-by: lubingtan <bingtanlu@gmail.com>

TerryHowe reviewed (approved):
/lgtm

gjenkins8 commented on pkg/chart/loader/load.go:
```diff
@@ -205,3 +206,48 @@ func LoadFiles(files []*BufferedFile) (*chart.Chart, error) {
 	return c, nil
 }
+
+// LoadValues loads chat values from a reader.
```
typo

And values are not necessary from a chart. They can be either be a chart's default values. Or user supplied values

LuBingtan replied:
thanks, updated here

LuBingtan committed:
fix: improve LoadValues function documentation

Signed-off-by: lubingtan <bingtanlu@gmail.com>

LuBingtan commented:
Hi @gjenkins8 would you please take a look again?

gjenkins8 reviewed (dismissed):
LGTM

robertsirc reviewed (dismissed):
LGTM

robertsirc commented:
Hi @LuBingtan your branch has some conflicts could you please take a look

LuBingtan committed:
fix: replace mergeMaps call with loader.MergeMaps in MergeValues function

Signed-off-by: lubingtan <bingtanlu@gmail.com>

gjenkins8 commented:
Thanks for the PR too!

Just to note, the `main` branch is currently targeting Helm v4. This change could be backported to Helm 3 if wanted. Via a PR to the `dev-v3` branch.

https://github.com/helm/helm/blob/main/CONTRIBUTING.md#helm-v3-and-v4 for more info

sabre1041 commented:
LGTM
